File: sdk/python/examples/langgraph/04_simple_stategraph.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Simple query pipeline — validate → refine → answer.

Demonstrates:
    - Using create_agent with a processing tool to build a pipeline
    - System prompt directing multi-step behavior (validate, refine, answer)
    - Server-side LLM orchestration (AI_MODEL task) + Python tool worker

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)

SYSTEM_PROMPT = (
    "You are a knowledgeable assistant that follows a two-step process:\n"
    "1. First, use the validate_query tool to clean and validate the user's question.\n"
    "2. Then answer the validated, refined question clearly and concisely.\n"
    "Always call validate_query before answering."
)


@tool
def validate_query(query: str) -> str:
    """Validate and clean the user's query. Returns the cleaned query or a default prompt."""
    cleaned = query.strip()
    if not cleaned:
        return "What can you help me with?"
    # Normalize punctuation
    if not cleaned.endswith("?") and not cleaned.endswith("."):
        cleaned = cleaned + "?"
    return cleaned


graph = create_agent(
    llm,
    tools=[validate_query],
    name="query_pipeline",
    system_prompt=SYSTEM_PROMPT,
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(graph, "Tell me about Python")
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/04-simple-stategraph.ts
/**
 * Simple StateGraph -- custom query → refine → answer pipeline.
 *
 * Demonstrates:
 *   - Defining a typed state schema with Annotation
 *   - Building a StateGraph with multiple sequential nodes
 *   - LLM calls inside node functions (detected by Agentspan for interception)
 *   - Connecting nodes with addEdge
 *   - Compiling and running via AgentRuntime
 *
 * Requirements:
 *   - AGENTSPAN_SERVER_URL=http://localhost:6767/api
 *   - OPENAI_API_KEY for ChatOpenAI
 */

import { StateGraph, START, END, Annotation } from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { AgentRuntime } from '@agentspan-ai/sdk';

const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });

// ---------------------------------------------------------------------------
// State schema
// ---------------------------------------------------------------------------
const QueryState = Annotation.Root({
  query: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  refined_query: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  answer: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type State = typeof QueryState.State;

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
function validate(state: State): Partial<State> {
  let query = (state.query || '').trim();
  if (query === '') {
    query = 'What can you help me with?';
  }
  return { query, refined_query: '', answer: '' };
}

async function refine(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('Rewrite the user query to be more specific and clear. Return only the rewritten query.'),
    new HumanMessage(state.query),
  ]);
  return { refined_query: (response.content as string).trim() };
}

async function answer(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('You are a knowledgeable assistant. Answer the question clearly and concisely.'),
    new HumanMessage(state.refined_query || state.query),
  ]);
  return { answer: (response.content as string).trim() };
}

// ---------------------------------------------------------------------------
// Build the graph
// ---------------------------------------------------------------------------
const builder = new StateGraph(QueryState);
builder.addNode('validate', validate);
builder.addNode('refine', refine);
builder.addNode('generate_answer', answer);
builder.addEdge(START, 'validate');
builder.addEdge('validate', 'refine');
builder.addEdge('refine', 'generate_answer');
builder.addEdge('generate_answer', END);

const graph = builder.compile({ name: "query_pipeline" });

// Add agentspan metadata for graph-structure extraction.
// NOTE: Do NOT set tools on StateGraphs — only model + framework.
(graph as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  framework: 'langgraph',
};

// ---------------------------------------------------------------------------
// Run on agentspan
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(graph, 'Tell me about Python');
    console.log('Status:', result.status);
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(graph);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples/langgraph --agents simple_stategraph
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(graph);
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
